src/object/obj_enum.c

```diff
@@ -889,6 +857,169 @@ dss_enum_unpack(vos_iter_type_t type, struct dss_enum_arg *arg,
+		       sizeof(daos_anchor_t));
+		memcpy(&oeda->akey_anchor, &oeda->akey_anchor_saved,
+		       sizeof(daos_anchor_t));
+		memcpy(&oeda->anchor, &oeda->anchor_saved,
+		       sizeof(daos_anchor_t));
+
+		D_ASSERT(oeda->size <= oeda->buf_len);
+
+		if (oeda->buf_len_saved < oeda->size) {
+			if (oeda->buf_saved != NULL)
+				D_FREE(oeda->buf_saved);
+
+			oeda->buf_len_saved = oeda->buf_len;
+			D_ALLOC(oeda->buf_saved, oeda->buf_len_saved);
+			if (oeda->buf_saved == NULL)
+				return -DER_NOMEM;
+		}
+
+		memcpy(oeda->buf_saved, oeda->buf, oeda->size);
+		oeda->size_saved = oeda->size;
+	} else {
+		/* Backup the anchors. */
+		memcpy(&oeda->dkey_anchor_saved, &oeda->dkey_anchor,
+		       sizeof(daos_anchor_t));
+		memcpy(&oeda->akey_anchor_saved, &oeda->akey_anchor,
+		       sizeof(daos_anchor_t));
+		memcpy(&oeda->anchor_saved, &oeda->anchor,
+		       sizeof(daos_anchor_t));
+
+		memset(oeda->kds, 0, sizeof(daos_key_desc_t) * KDS_NUM);
+		memset(oeda->eprs, 0, sizeof(daos_epoch_range_t) * KDS_NUM);
+		memset(oeda->buf, 0, oeda->buf_len);
+
+		oeda->iov.iov_len = 0;
+		oeda->iov.iov_buf = oeda->buf;
+		oeda->iov.iov_buf_len = oeda->buf_len;
+
+		oeda->sgl.sg_nr = 1;
+		oeda->sgl.sg_nr_out = 1;
+
+		oeda->size = 0;
+		oeda->num = KDS_NUM;
+
+		oeda->has_retried = 0;
+		oeda->lost_shard = 0;
+	}
+
+	daos_anchor_set_flags(&oeda->dkey_anchor, flags);
+
+	rc = list_cb(oh, epoch, NULL, NULL, &oeda->size, &oeda->num, oeda->kds,
+		     oeda->eprs, &oeda->sgl, &oeda->anchor, &oeda->dkey_anchor,
+		     &oeda->akey_anchor);
+	if (rc == -DER_KEY2BIG) {
+		D_DEBUG(DB_REBUILD, "list obj dkeys on shard "DF_UOID
+			" got -DER_KEY2BIG, key_len "DF_U64"\n",
+			DP_UOID(oeda->enum_arg.oid), oeda->kds[0].kd_key_len);
+		oeda->buf_len = roundup(oeda->kds[0].kd_key_len * 2, 8);
+		if (oeda->buf != oeda->inline_buf)
+			D_FREE(oeda->buf);
+
+		D_ALLOC(oeda->buf, oeda->buf_len);
+		if (oeda->buf == NULL)
+			rc = -DER_NOMEM;
+	}
+
+	return rc;
+}
+
+int
+daos_enum_dkeys(daos_handle_t oh, daos_unit_oid_t oid, daos_epoch_t epoch,
+		daos_obj_list_obj_cb_t list_cb,
+		daos_enum_unpack_cb_t unpack_cb, void *arg)
+{
+	struct obj_enum_dkeys_arg	oeda;
+	int				rc = 0;
+
+	daos_enum_dkeys_init_arg(&oeda, oid);
+
+	do {
+		rc = daos_enum_dkeys_do_list(oh, &epoch, &oeda, list_cb,
+					     DIOF_TO_LEADER);
+		/* Enlarge the buffer and re-list the dkeys. */
+		if (rc == -DER_KEY2BIG)
+			continue;
+
+		if (rc != 0) {
+			/* container might have been destroyed. Or there is
+			 * no spare target left for this object see
+			 * obj_grp_valid_shard_get()
+			 */
+			if (rc == -DER_NONEXIST)
+				rc = 0;
+
+			break;
+		}
+
+		if (oeda.num == 0)
+			break;
+
+		daos_enum_dkeys_prep_unpack(&oeda);
+		rc = daos_enum_unpack(VOS_ITER_DKEY, &oeda.enum_arg,
+				      unpack_cb, arg);
+		if (rc != 0) {
+			D_ERROR("list obj dkeys "DF_UOID" failed: rc = %d\n",
+				DP_UOID(oid), rc);
+			break;
+		}
+	} while (!daos_anchor_is_eof(&oeda.dkey_anchor));
```

Comment:
I would suggest not, because the dc_obj_verify_replicas() has more logic to be handled than object rebuild case, such as fetch multiple replicas, retry the fetch if detect inconsistency for potential DTX commit, compare replicas, and so on. If we share the logic with object rebuild, most of the logic is meaningless for the latter one.